content/en/blog/_posts/2021-08-11-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,88 @@
+
+## How does it work?
+It uses two interfaces, `memory.min` and `memory.high` of cgroup v2 memory controller, to work.
+
+<table>
+    <tr>
+        <th style="text-align:center">File</th>
+        <th style="text-align:center">Description</th>
+   </tr>
+   <tr>
+        <td>memory.min</td>
+        <td><code>memory.min</code> specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked.
+        <br>
+        <br>
+        <i>We map it to <code>requests.memory</code></i>
+        </td>
+   </tr>
+   <tr>
+       <td>memory.high</td>
+       <td><code>memory.high</code> is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. 
+       <br>
+       <br>
+       <i>We use a formula to calculate <code>memory.high</code>, depending on <code>limits.memory/node allocatable memory</code> and a memory throttling factor</i>
+       </td>
+   </tr>
+</table>
+
+Both containers and pods are considered. When cgroup v2 unified and MemoryQoS feature gate are both enabled, kubelet sets `requests.memory` to `memory.min` at container and pod level cgroup. And it also sets `memory.high` to throttle container memory overcommit allocation.
+
+![](./memory-qos-cal.svg)
+
+The formula is as follows:  
+1、Container    
+![](./container-memory-min.svg)
+![](./container-memory-high.svg)
+
+2、Pod  
+![](./pod-memory-min.svg)
+
+3、QoS  
+![](./qos-memory-min.svg)
+
+
+The cgroup v2 memory controller enforces memory requests and ceilings per-pod or per-container. If you set. The design of cgroup v2 also implies for the `memory.min` behavior to be as described, all ancestor cgroups must have appropriate `memory.min` values. To ensure this, the kubelet sets `memory.min` on all ancestor cgroups up to the root. You are responsible for ensuring that your init system (or other operating system level tooling) does not also try to manage `memory.min` on those ancestor cgroups.
```

Comment:
Again, kubelet doesn't set this on the cgroup directly, the kubelet assigns values though.